### DISCUSSION TITLE
Is there a way to register a custom extension?

### DISCUSSION THREAD

**Discussion by Niklan** [phpactor/phpactor#2392] [Q&A] [answered]
Hi, I'm interesting of trying to create a custom extension for a specific integration. I see there is an extension system, but I couldn't find any public API for it. The only way to register a custom extension looks like through override of `container.extension_classes` which will require to list default ones as well.

But I'm not sure if it would work, because they are hard-coded and then registered: 

https://github.com/phpactor/phpactor/blob/dc34cf282daebb08d7ad78a855492ebebaf02e99/lib/Phpactor.php#L126-L130

I saw there is a mention here (https://github.com/phpactor/phpactor/discussions/2365) that extending an indexer is not supported, but I don't think I need that, I want to create more like Symfony extension, but just for Drupal (because services defined differently and Symfony extension can't handle it).

Is that possible at this moment at all, without patching or hacking, or not? If not, is this feature planned?

**Comment by dantleech** [comment, marked as answer]
I made the decision a while back to remove the "extension manager" that previously supported user extensions:

1) it was based on composer 1.0 and needed to be written from scratch
2) there were not any serious extensions in the wild not made by me
3) Phpactor's API was an still is highly unstable. I don't want to preserve/worry about B/C for the sake of extensions.
4) Not sure how "extensions" (which may have their own dependencies) would work with the PHAR distribution.

So, all extensions are now bundled in the Phpactor  "mono repo". There a limitations and benefits to this. One benefit is that I can better guarantee that all extensions are supported and will continue to work

If you open an issue or start a discussion we can see if your extension could be contributed to Phpactor.

**Reply by Niklan** [reply]
I'll try to create an extension via patch for phpactor itself. It's okay for me, since I just want to play with it, if it will end up with something useful, it will be more sense to start a discussion or provide a PR.

Now, it's clears for me that there is no public API for custom extensions. Because the code looks like designed for that, but the entry point simply hard-coded them, and I was confused :smile: 

Thank you.
